compiler/luci-interpreter/include/luci_interpreter/BuddyMemoryManager.h

```diff
@@ -0,0 +1,141 @@
+/* Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci_interpreter/MemoryManager.h"
+
+#ifndef LUCI_INTERPRETER_BUDDY_MEMORY_MANAGER_H
+#define LUCI_INTERPRETER_BUDDY_MEMORY_MANAGER_H
+
+namespace luci_interpreter
+{
+
+class BuddyMemoryManager : public IMemoryManager
+{
+public:
+  BuddyMemoryManager(uint8_t *memory_start, int32_t memSize);
+
+  void allocate_memory(luci_interpreter::Tensor &tensor) final;
+  void release_memory(luci_interpreter::Tensor &tensor) final;
+
+private:
+  struct Block
+  {
+    Block *next_free;
+    Block *self;
+    bool is_free;
+    uint32_t size;
+  };
+
+  Block *_start_block;
+  int32_t _num_blocks;
+  uint32_t _size;
+  Block *_free_blocks[32]{};
+
+  static int32_t powOf2(uint32_t val)
```

Comment:
Better to name this function `upper_log2` because this func returns `(val - 1) & val == 0 ? int(log2(val)) : int(log2(val)) + 1`.

Reply:
@m-bronnikov interesting article, by the way =)

Reply:
Done

Reply:
> upper_log2

If I understand what this function is doing, it is **lower**_log2

Reply:
> > upper_log2
> 
> If I understand what this function is doing, it is more look like **lower**_log2

Yes, you are right, thank you!